Apollo/Devices/Loop.cs

```diff
@@ -0,0 +1,113 @@
+                if (_duration != null) {
+                    _duration.FreeChanged -= FreeChanged;
+                    _duration.ModeChanged -= ModeChanged;
+                    _duration.StepChanged -= StepChanged;
+                }
+
+                _duration = value;
+
+                if (_duration != null) {
+                    _duration.Minimum = 10;
+                    _duration.Maximum = 30000;
+
+                    _duration.FreeChanged += FreeChanged;
+                    _duration.ModeChanged += ModeChanged;
+                    _duration.StepChanged += StepChanged;
+                }
+            }
+        }
+        
+        int _amount = 0;
+        public int Amount {
+            get => _amount;
+            set {
+                if (0 <= value && value <= 100 && _amount != value) {
+                    _amount = value;
+                    
+                    if (Viewer?.SpecificViewer != null) ((LoopViewer)Viewer.SpecificViewer).SetAmount(Amount);
+                }
+            }
+        }
+        
+        ConcurrentDictionary<Signal, Courier> timers = new ConcurrentDictionary<Signal, Courier>();
+
+        void FreeChanged(int value) {
+            if (Viewer?.SpecificViewer != null) ((LoopViewer)Viewer.SpecificViewer).SetDurationValue(value);
+        }
+
+        void ModeChanged(bool value) {
+            if (Viewer?.SpecificViewer != null) ((LoopViewer)Viewer.SpecificViewer).SetMode(value);
+        }
+
+        void StepChanged(Length value) {
+            if (Viewer?.SpecificViewer != null) ((LoopViewer)Viewer.SpecificViewer).SetDurationStep(value);
+        }
+        
+        public Loop(Time duration = null, int amount = 0) : base("loop"){
+            Duration = duration?? new Time();
+            Amount = amount;
+        }
+        
+        public override Device Clone() => new Loop(Duration, Amount);
+        
+        public override void MIDIProcess(Signal n){
+            Courier courier;
+            timers[n] = courier = new Courier(){
+                Info = new LoopCourierInfo(0, Amount, n),
+                Interval = Duration,
+                AutoReset = true
+            };
+            
+            courier.Elapsed += Tick;
+            courier.Start();
+            
+            Signal m = n.Clone();
+            InvokeExit(m);
+        }
+        
+        void Tick(object sender, EventArgs e) {
+            Courier courier = (Courier)sender;
+            LoopCourierInfo info = (LoopCourierInfo) courier.Info;
+            
+            int count = info.count + 1;
+            
+            if(count > info.amount){
+                courier.Stop();
+                courier.Elapsed -= Tick;
+                courier.Dispose();
+                return;
+            }
+            
+            info.count = count;
+            
+            timers[info.signal].Info = info;
+                        
+            Signal m = info.signal.Clone();
+            InvokeExit(m);
+        }
+    }
```

Comment:
Missing Dispose function for the timers?

Reply:
This is done on line 101, but I'm changing the way the timers work anyway so this will be outdated

Reply:
You're still missing a Dispose, which should run when the Device is deleted, or parent disposes itself. See how Fade, Delay, etc disposes are done

Reply:
I've overrided Dispose() and Stop() in a similar way to the other devices, is it enough?
```c#
protected override void Stop() {
    foreach (List<Courier> i in timers.Values) {
        foreach (Courier j in i) j.Dispose();
            i.Clear();
    }
    timers.Clear();
}
        
public override void Dispose() {
    if (Disposed) return;
            
    Stop();

    base.Dispose();
}
```

Reply:
I guess push it so i can poke around